LanguageFeatures/Patterns/type_inference_A02_t01.dart

```diff
@@ -17,38 +17,9 @@
 
 import "../../Utils/static_type_helper.dart";
 
-class A {}
-class B extends A {}
-class C extends B {}
-class D extends C {}
-
 main() {
-  A? a = A();
-  B? b = B();
-  C? c = C();
-  D? d = D();
-  {
-    var ([A? v1, B? v2, v3]!) = [b, c, c];
-    v1.expectStaticType<Exactly<A>>();
-    v2.expectStaticType<Exactly<B>>();
-    v3.expectStaticType<Exactly<C>>();
-  }
-  {
-    final ([A? v1, B? v2, v3]!) = [b, c, d];
-    v1.expectStaticType<Exactly<A>>();
-    v2.expectStaticType<Exactly<B>>();
-    v3.expectStaticType<Exactly<D>>();
-  }
-  {
-    var ([A? v1, B? v2, v3]!) = <C?>[d, d, d];
-    v1.expectStaticType<Exactly<A>>();
-    v2.expectStaticType<Exactly<B>>();
-    v3.expectStaticType<Exactly<C>>();
-  }
-  try {
-    final ([A? v1, B? v2, v3]!) = [];
-    v1.expectStaticType<Exactly<A>>();
-    v2.expectStaticType<Exactly<B>>();
-    v3.expectStaticType<Exactly<B>>();
-  } catch (_) {}
+  var (v1!) = 42 as num?;
+  v1.expectStaticType<Exactly<num>>();
+  final (v2!) = 42 as num?;
+  v2.expectStaticType<Exactly<A>>();
```

Comment:
This is tricky, as usual.

Assume `X whatever<X extends Object?>(Object? o) { ... return o as X; }` where `...` will do some logging of the value of `X` (and we avoid worrying about inference yielding `dynamic` when there's no bound, because there isn't no bound ;-).

If we are testing `var (num v1!) = whatever(1);` then the actual type argument will reflect the context type schema (it's the greatest closure), so `X` should get the value `num?`.

I think the tests in line 21 and 23 are less safe, because there's no way for the context type schema to influence the typing of the initializing expression, which means that the declared type of `v1` is decided based on the _third_ step of the static analysis: Type checking and finalizing the pattern, based on the static type of the initializing expression.

It is not even an error for type inference to fail (say, because the context type schema is very different from the static type of the expression under inference), and I believe that it would be possible for the context type schema of `(v1!)` to be `Map<Widget, _>` or basically anything else, and we would still see the same behavior.

So we really need to have a testing set up where the initializing expression supports exploration of the context type schema, rather than a test where the context type schema is basically blocked from having any effect at all.

A function like `whatever` can do that, and the fact that it takes an argument makes it possible for the invocation of `whatever` to complete normally, which means that it's OK to have it in a run-time test. (Surely we can `try`/`catch`, but this is more convenient.)

The next step is that we'd need to test context type schemas where the unknown type `_` occurs in one or more locations. That's even more tricky. So how about aiming for a set of tests where the context type schema has no occurrences of `_`, and then we can see that the actual static analysis behavior matches the specification in those cases.

Reply:
I'd expect that we have a bunch of cases of the general type 'Check that the context type schema for a ... pattern is a context type schema ... where ...'.

They will all fit nicely with this kind of test where we use `whatever` to get hold of `X` from a context type schema which is based on a declared type in a variablePattern, plus some extra stuff around the variablePattern which will change the context type schema in a specified manner.

Reply:
I can see that lines 21-24 were changed. As I mentioned, we can hardly claim that we're testing the computation of the type schema of a pattern in the case where the initializing expression type inference will be able to simply ignore the given context type schema.

So what are we actually testing here?

I think we're testing the 3rd phase, type checking and finalization of the pattern. So we have an initializing expression whose type is `num?`, then we're type checking `(v1!)` with matched value type `num?`, i.e., `v1!` with `num?`, i.e., `v1` with `num`, and the result is that `v1` gets the inferred declared type `num`.

I think the `@description` should be adjusted to say that we're testing the type checking phase of patterns, or something like that.

Line 24 should be adjusted to expect type `num` as well.